Allow running multiple SSH daemons

SteffenDE opened the pull request:
Hello there,

while creating [nerves_ssh_shell](https://github.com/SteffenDE/nerves_ssh_shell) to connect to regular shells on nerves devices, I found the need to start multiple ssh daemons (as ssh subsystems do not allow the same set of features than running a separate daemon). As I also did not want to reimplement what nerves_ssh is already doing very well, I thought that adding the possibility to run multiple daemons seems like something that might benefit others too.

This is not complete yet, for example I only changed the tests to pass for now and did not add new tests for starting multiple daemons, as I first of all wanted some feedback what you all think about this. I also added a configuration option for the used `ssh_cli` setting, but this might not be needed as there is already the `daemon_option_overrides`.

SteffenDE commented:
Yes, I think that supporting Unix shells in nerves_ssh would be very cool and I'd be very happy to work on this being available by default.

For this pull request, I think the focus should still be making running multiple daemons possible first, as this is the basis for supporting other shells running next to IEx when not using subsystems (that have some limitations as documented in the `:nerves_ssh_shell` README).

I think it probably makes sense to continue the discussion around unix shell support, so I've opened #84 for that.

jjcarstens commented on lib/nerves_ssh/options.ex:
```diff
@@ -228,6 +231,9 @@ defmodule NervesSSH.Options do
   defp shell_opts(%{shell: :disabled}), do: [shell: :disabled]
 
+  defp cli_opts(%{cli: nil}), do: []
+  defp cli_opts(%{cli: cli}), do: [ssh_cli: cli]
```
I think that I would vote we leave the options.ex file alone for now. If a custom shell is being used, we just supply the options in `daemon_option_overrides: [ssh_cli: opts]`

jjcarstens commented on lib/nerves_ssh.ex:
```diff
@@ -26,18 +26,24 @@ defmodule NervesSSH do
     defstruct opts: [], sshd: nil, sshd_ref: nil
   end
 
+  defp via_tuple(name), do: {:via, Registry, {NervesSSH.Registry, name}}
```
```suggestion
  defp via_name(name), do: {:via, Registry, {NervesSSH.Registry, name}}
```
This is typed as a `GenServer.name()`, so I typically call these `via_name/1` functions.

SteffenDE replied:
Sounds reasonable (I got the name `via_tuple` from Saša Jurić's Elixir in Action)

jjcarstens reviewed (commented):
Just a couple comments, but otherwise seems good to me

SteffenDE commented:
So, I just added one test for actually starting two daemons and found an issue with `NervesSSH.UserPasswords` and `NervesSSH.Keys` needing access to the daemon name. I changed the code accordingly and passed the registered name to the `NervesSSH.Options` struct.

jjcarstens commented on lib/nerves_ssh.ex:
```diff
@@ -26,18 +26,24 @@ defmodule NervesSSH do
   @doc false
-  @spec start_link(Options.t()) :: GenServer.on_start()
+  @spec start_link(Options.t() | {atom(), Options.t()}) :: GenServer.on_start()
+  def start_link({name, %Options{} = opts}) do
```
I think since we now need `:name` as part of the options, we should remove this start_link function header and only rely on the one below expecting the options struct.

jjcarstens commented on lib/nerves_ssh.ex:
```diff
@@ -26,18 +26,24 @@ defmodule NervesSSH do
+
   def start_link(%Options{} = opts) do
-    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
+    GenServer.start_link(__MODULE__, %{opts | name: :default}, name: via_name(:default))
```
```suggestion
GenServer.start_link(__MODULE__, opts, name: via_name(opts.name))
```
Let's expect options to handle the default name

jjcarstens commented on lib/nerves_ssh/options.ex:
```diff
@@ -40,7 +41,8 @@ defmodule NervesSSH.Options do
         }
 
-  defstruct authorized_keys: [],
+  defstruct name: :default,
```
I'm going back and forth on having the default name just continue to be `NervesSSH` rather than `:default` here. It seems like that would be much less of an interrupt than using a new atom of `:default` since that has been the expected name so far for the life of this app. What do you think?

SteffenDE replied:
I think that's fine, done 👍

fhunleth reviewed (approved):
LGTM!
